Fix c10::util::get_fully_qualified_type_name for MSVC

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26482 [wip] Error checking for KernelFunction::callUnboxed()
* #30712 Better error messages thanks to compile time type names
* #31307 [no-land] try fix 2
* #31306 [no-land] try fix 1
* #26628 Remove CAFFE_KNOWN_TYPE
* **#31313 Fix c10::util::get_fully_qualified_type_name for MSVC**
* #31351 Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4
* #31312 constexpr typeid improvements

This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately.

Proof that this works: The `C10_TYPENAME_SUPPORTS_CONSTEXPR` macro was previously off for msvc because test cases enable additional compile-time tests based on that macro and those failed for msvc. Now, this macro is enabled for msvc and the tests pass.

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

smessmer committed:
Fix c10::util::get_fully_qualified_type_name for MSVC

This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately...

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 172275a:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=172275a446640565e34a614ed6f6236b4209810b).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_windows_build</sup> | <sup>Build</sup> | New in PR |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "Fix c10::util::get_fully_qualified_type_name for MSVC"


This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately.

Proof that this works: The `C10_TYPENAME_SUPPORTS_CONSTEXPR` macro was previously off for msvc because test cases enable additional compile-time tests based on that macro and those failed for msvc. Now, this macro is enabled for msvc and the tests pass.

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix c10::util::get_fully_qualified_type_name for MSVC"


This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately.

Proof that this works: The `C10_TYPENAME_SUPPORTS_CONSTEXPR` macro was previously off for msvc because test cases enable additional compile-time tests based on that macro and those failed for msvc. Now, this macro is enabled for msvc and the tests pass.

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix c10::util::get_fully_qualified_type_name for MSVC"


This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately.

Proof that this works: The `C10_TYPENAME_SUPPORTS_CONSTEXPR` macro was previously off for msvc because test cases enable additional compile-time tests based on that macro and those failed for msvc. Now, this macro is enabled for msvc and the tests pass.

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
If it passes CI and is tested - let's ship it

smessmer committed:
Update on "Fix c10::util::get_fully_qualified_type_name for MSVC"


This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately.

Proof that this works: The `C10_TYPENAME_SUPPORTS_CONSTEXPR` macro was previously off for msvc because test cases enable additional compile-time tests based on that macro and those failed for msvc. Now, this macro is enabled for msvc and the tests pass.

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix c10::util::get_fully_qualified_type_name for MSVC"


This is a bugfix. The reason we couldn't enable the constexpr-ness for it before is that it was buggy,
and without constexpr it crashed at runtime and not at compile time which seems to have passed our CI unfortunately.

Proof that this works: The `C10_TYPENAME_SUPPORTS_CONSTEXPR` macro was previously off for msvc because test cases enable additional compile-time tests based on that macro and those failed for msvc. Now, this macro is enabled for msvc and the tests pass.

Differential Revision: [D19087471](https://our.internmc.facebook.com/intern/diff/D19087471/)

[ghstack-poisoned]
